docs/versioning-soroban.md

```diff
@@ -0,0 +1,148 @@
+     when configured with two versions of soroban built-in, and such a
+     configuration will cut over instantaneously from old soroban `N.x.y` to new
+     soroban `(N+1).0.0` on upgrade.
+
+  6. Once the upgrade is complete, we prepare the _next_ version of stellar-core
+     _without_ the transitionary version. That is, we build without
+     `--feature=soroban-env-host-prev`. In order to be sure that it is safe to
+     expire that version, we first replay all ledgers from `N` and before on
+     the current soroban `(N+1).0.0` and see if they replay correctly. If they
+     do not, we add backward compatibility code to soroban and do a point
+     release like `(N+1).0.1` that can replay correctly, and bump the version
+     in stellar-core's `Cargo.toml` to `(N+1).0.1`.
+
+## Pinned soroban dependency trees
+
+At any point we may have one _or two_ versions of soroban built into
+stellar-core (see below). This is confusing and error-prone. It is easy to
+update one version and accidentally cause cargo to "re-resolve" version specs in
+dependencies or transitive dependencies to new versions, and accidentally cause
+changes.
+
+To help cross-check versions and reduce accidental changes, we embed (in all
+builds) a copy of both `Cargo.lock` and one or more files showing the full,
+expected dependency tree of either or both of the embedded copies of
+soroban. These files are stored in git adjacent to `Cargo.toml` and are called
+`host-dep-tree-curr.txt` and `host-dep-tree-prev.txt`. Any time we change the
+version(s) specified in `Cargo.toml` we will likely need to update one or
+another of these files.
+
+The easiest way to update the files is just to run `stellar-core version`
+without updating the files: it will exit with an error that shows the file
+content that was expected and the content that was found. Copy the "found"
+content into the expected file and rebuild, and it should work. The point of
+this step is to catch accidental errors, so we carefully review the changed
+dependency tree before we commit it, and ensure it contains exactly the changes
+we expect!
+
+If we want to manually regenerate the files, we can also use the cli tool
+`cargo-lock` by hand (this is the crate stellar-core uses), for example:
+
+    $ cargo install --git https://github.com/rustsec/rustsec \
+        --rev dcc72b697e10a2d1e3d1ce70d7d7b0d5cbc41dc4 \
+	--locked --features cli cargo-lock
+    $ cargo lock tree --exact soroban-env-host@0.0.15
+
+
+## Protocol version, release version and contract cross-checks
+
+We configure soroban, stellar-core and contracts built for soroban to check
+certain consistency requirements around protocol and release versions:
+
+  1. Stellar-core's max protocol (`Config:::CURRENT_LEDGER_PROTOCOL_VERSION`)
+     and its major version number must be the same.
+
+  2. Stellar-core's and soroban's major version numbers are the same.
+
+  3. Every contract's WASM has an embedded minimum protocol version number,
```

Comment:
Can we reuse the env meta version for this? It seems like it has to have 1:1 mapping with the protocol versions and also it's easier to notice the mismatch when using e.g. CLI.